pkg/controller/sync/controller.go

```diff
@@ -373,3 +364,123 @@ func (s *FederationSyncController) syncToClusters(fedResource FederatedResource,
+
+	finalizers := sets.NewString(obj.GetFinalizers()...)
+	if !finalizers.Has(FinalizerSyncController) {
+		glog.V(2).Infof("%s %q does not have the %q finalizer. Nothing to do.", kind, key, FinalizerSyncController)
+		return nil
+	}
+
+	annotations := obj.GetAnnotations()
+	orphanResources := annotations != nil && annotations[OrphanManagedResources] == "true"
+	if orphanResources {
+		glog.V(2).Infof("Found %q annotation on %s %q. Removing the finalizer.", OrphanManagedResources, kind, key)
+		err := s.removeFinalizer(fedResource)
+		if err != nil {
+			return errors.Wrapf(err, "failed to remove finalizer %q", OrphanManagedResources)
+		}
+		glog.V(2).Infof("Initiating the removal of the label %q from resources previously managed by %s %q.", util.ManagedByFederationLabelKey, kind, key)
+		return nil
+		// TODO(marun) Implement removal of labels
+	}
+
+	glog.V(2).Infof("Deleting resources managed by %s %q from member clusters.", kind, key)
+
+	// First retrieve a list of unready clusters to maximize the
+	// changes of detecting when it's not possible to guarantee
+	// removal from member clusters.
+	unreadyClusters, err := s.informer.GetUnreadyClusters()
+	if err != nil {
+		return errors.Wrap(err, "failed to get a list of unready clusters")
+	}
+
+	clusters, err := s.informer.GetReadyClusters()
+	if err != nil {
+		return errors.Wrap(err, "failed to get a list of ready clusters")
+	}
+
+	updater := fedResource.NewUpdater()
+	deletedFromClusterNames := []string{}
+	targetKey := fedResource.TargetName().String()
+	retrievalErrorClusterNames := []string{}
+	for _, cluster := range clusters {
+		clusterName := cluster.Name
+		rawClusterObj, _, err := s.informer.GetTargetStore().GetByKey(clusterName, targetKey)
+		if err != nil {
+			runtime.HandleError(errors.Wrapf(err, "failed to retrieve managed resource for cluster %q", clusterName))
+			retrievalErrorClusterNames = append(retrievalErrorClusterNames, clusterName)
+			continue
+		}
+		// Already removed
+		if rawClusterObj == nil {
+			continue
+		}
+
+		clusterObj := rawClusterObj.(*unstructured.Unstructured)
+		if fedResource.IsNamespaceInHostCluster(clusterObj) {
+			// Host cluster namespace needs to have the managed
+			// label removed so it won't be cached anymore.
+			updater.RemoveManagedLabel(clusterName, clusterObj)
+		} else {
+			updater.Delete(clusterName)
+			deletedFromClusterNames = append(deletedFromClusterNames, clusterName)
+		}
+	}
+	if updater.NoChanges() {
+		if len(retrievalErrorClusterNames) > 0 {
+			return errors.Errorf("Failed to retrieve a managed resource from the following cluster(s): %v", strings.Join(retrievalErrorClusterNames, ", "))
+		}
+
+		if len(unreadyClusters) == 0 {
+			// Managed resources no longer exist in any member cluster.
+			return s.removeFinalizer(fedResource)
+		}
+
+		// It is not possible to know if resources have been removed
+		// from all clusters if one or more clusters is not ready.
+		var unreadyClusterNames []string
```

Comment:
i guess unreadyClusters check is also needed for !NoChanges case

Reply:
I've rewritten the logic such that this comment should be resolved.